README.rst
Djangorecipe: easy install of Django with buildout
==================================================


Description
-----------

.. image:: https://secure.travis-ci.org/rvanlaar/djangorecipe.png?branch=master
   :target: http://travis-ci.org/rvanlaar/djangorecipe/

This buildout recipe can be used to create a setup for Django. It will
automatically download Django and install it in the buildout's
sandbox.

You can see an example of how to use the recipe below::

  [buildout]
  parts = satchmo django
  eggs =
    ipython
  versions = versions

  [versions]
  Django = 1.5.5

  [satchmo]
  recipe = gocept.download
  url = http://www.satchmoproject.com/snapshots/satchmo-0.6.tar.gz
  md5sum = 659a4845c1c731be5cfe29bfcc5d14b1

  [django]
  recipe = djangorecipe
  settings = development
  eggs = ${buildout:eggs}
  extra-paths =
    ${satchmo:location}
  project = dummyshop


Supported options
-----------------

The recipe supports the following options.

project
  This option sets the name for your project. The recipe will create a
  basic structure if the project settings module does not already exist.
  If this option is set to '.', the project root is understood to be the
  buildout root.

projectegg
  Use this instead of the project option when you want to use an egg
  as the project. This disables the generation of the project
  structure.

settings
  You can set the name of the settings file which is to be used with
  this option. This is useful if you want to have a different
  production setup from your development setup. It defaults to
  `development`.

template
  The name of the template folder that should be used when creating a new
  project. By default, the recipe creates a django project exactly like the
  django-admin.py startproject command, matching the major django version you
  are using. For this option to be taken into account, the
  {buildout:django_template_dirs} must contain at least one entry.

extra-paths
  All paths specified here will be used to extend the default Python
  path for the `bin/*` scripts.

pth-files
  Adds paths found from a site `.pth` file to the extra-paths.
  Useful for things like Pinax which maintains its own external_libs dir.

control-script
  The name of the script created in the bin folder. This script is the
  equivalent of the `manage.py` Django normally creates. By default it
  uses the name of the section (the part between the `[ ]`).

initialization
  Specify some Python initialization code to be inserted into the
  `control-script`. This is very limited. In particular, be aware that
  leading whitespace is stripped from the code given.

wsgi
  An extra script is generated in the bin folder when this is set to
  `true`. This can be used with mod_wsgi to deploy the project. The
  name of the script is `control-script.wsgi`.

wsgi-script
  The name of the wsgi-script that is generated. This can be useful for
  gunicorn.

wsgilog
  In case the WSGI server you're using does not allow printing to stdout,
  you can set this variable to a filesystem path - all stdout/stderr data
  is redirected to the log instead of printed

test
  If you want a script in the bin folder to run all the tests for a
  specific set of apps this is the option you would use. Set this to
  the list of app labels which you want to be tested.

testrunner
  This is the name of the testrunner which will be created. It
  defaults to `test`.

All following options only have effect when the project specified by
the project option has not been created already.

urlconf
  You can set this to a specific url conf. It will use project.urls by
  default.

secret
  The secret to use for the `settings.py`, it generates a random
  string by default.


Dedicated buildout options
--------------------------

The recipe can optionally use an option from the buildout section to get the
directories where template folders lie.

django_template_dirs
   This is a list of directories where project template folders can be found
   The template is defined by the above option 'template'


Another example
---------------

The next example shows you how to use some more of the options::

  [buildout]
  parts = django extras
  eggs =
    hashlib

  [extras]
  recipe = iw.recipe.subversion
  urls =
    http://django-command-extensions.googlecode.com/svn/trunk/ django-command-extensions
    http://django-mptt.googlecode.com/svn/trunk/ django-mptt

  [django]
  recipe = djangorecipe
  settings = development
  project = exampleproject
  wsgi = true
  eggs =
    ${buildout:eggs}
  test =
    someapp
    anotherapp

Example using .pth files
------------------------

Pinax uses a .pth file to add a bunch of libraries to its path; we can
specify it's directory to get the libraries it specified added to our
path::

  [buildout]
  parts	= PIL
	  svncode
	  myproject
  versions=versions

  [versions]
  django	= 1.3

  [PIL]
  recipe	= zc.recipe.egg:custom
  egg		= PIL
  find-links	= http://dist.repoze.org/

  [svncode]
  recipe	= iw.recipe.subversion
  urls		= http://svn.pinaxproject.com/pinax/tags/0.5.1rc1	pinax

  [myproject]
  recipe	= djangorecipe
  eggs		=
    PIL
  project	= myproject
  settings	= settings
  extra-paths	= ${buildout:directory}/myproject/apps
		  ${svncode:location}/pinax/apps/external_apps
		  ${svncode:location}/pinax/apps/local_apps
  pth-files	= ${svncode:location}/pinax/libs/external_libs
  wsgi		= true

Above, we use stock Pinax for pth-files and extra-paths paths for
apps, and our own project for the path that will be found first in the
list.  Note that we expect our project to be checked out (e.g., by
svn:external) directly under this directory in to 'myproject'.


Example with a Django version from a repository
-----------------------------------------------

If you want to use a specific Django version from a source
repository you could use mr.developer: http://pypi.python.org/pypi/mr.developer
Here is an example for using the Django development version::

  [buildout]
  parts = django
  extensions = mr.developer
  auto-checkout = *

  [sources]
  django = git https://github.com/django/django.git

  [django]
  recipe = djangorecipe
  settings = settings
  project = project

Example configuration for mod_wsgi
----------------------------------

If you want to deploy a project using mod_wsgi you could use this
example as a starting point::

  <Directory /path/to/buildout>
         Order deny,allow
         Allow from all
  </Directory>
  <VirtualHost 1.2.3.4:80>
         ServerName      my.rocking.server
         CustomLog       /var/log/apache2/my.rocking.server/access.log combined
         ErrorLog        /var/log/apache2/my.rocking.server/error.log
         WSGIScriptAlias / /path/to/buildout/bin/django.wsgi
  </VirtualHost>

Generating a control script for PyDev
-------------------------------------

Running Django with auto-reload in PyDev requires adding a small snippet
of code::

  import pydevd
  pydevd.patch_django_autoreload(patch_remote_debugger=False, patch_show_console=True)

just before the `if __name__ == "__main__":` in the `manage.py` module
(or in this case the control script that is generated). This example
buildout generates two control scripts: one for command-line usage and
one for PyDev, with the required snippet, using the recipe's
`initialization` option::

  [buildout]
  parts = django pydev
  eggs =
    mock

  [django]
  recipe = djangorecipe
  eggs = ${buildout:eggs}
  project = dummyshop

  [pydev]
  <= django
  initialization =
    import pydevd
    pydevd.patch_django_autoreload(patch_remote_debugger=False, patch_show_console=True)

Several wsgi scripts for one Apache virtual host instance
---------------------------------------------------------

There is a problem when several wsgi scripts are combined in a single virtual
host instance of Apache. This is due to the fact that Django uses the
environment variable DJANGO_SETTINGS_MODULE. This variable  gets set once when
the first wsgi script loads. The rest of the wsgi scripts will fail, because
they need a different settings modules. However the environment variable
DJANGO_SETTINGS_MODULE is only set once. The new `initialization` option that has
been added to djangorecipe can be used to remedy this problem as shown below::

    [django]
    settings = acceptance
    initialization =
        import os
        os.environ['DJANGO_SETTINGS_MODULE'] = '${django:project}.${django:settings}'

Example usage of django-configurations
--------------------------------------

django-configurations (http://django-configurations.readthedocs.org/en/latest/)
is an application that helps you organize your Django settings into classes.
Using it requires modifying the manage.py file.  This is done easily using the
recipe's `initialization` option::

    [buildout]
    parts = django
    eggs =
        hashlib

    [django]
    recipe = djangorecipe
    eggs = ${buildout:eggs}
    project = myproject
    initialization =
        # Patch the manage file for django-configurations
        import os
        os.environ.setdefault('DJANGO_SETTINGS_MODULE', 'myproject.settings')
        os.environ.setdefault('DJANGO_CONFIGURATION', 'Development')
        from configurations.management import execute_from_command_line
        import django
        django.core.management.execute_from_command_line = execute_from_command_line


Templating
----------

The template engine is as simple as it can be and relies upon pythons's
string.Template. A variable can be inserted in any file or directory name or
file content using the syntax ${variable}.

The following variables are available:

- all the recipe options from the configuration file
- secret: the secret key for django settings
- project_name: the project name (project or the buildout directory name if
  project == '.')
- root_pkg: the root package (empty string if project == '.')
- year: the current year
- month: the current month
- day: the current day of the month

You may use these variables in any file of the template directory.

For example, for a copyright notice in a module's docstring, you may use::

   (c) ${year} Me

Example:
........

In ~/.buildout/default.cfg::

    [buildout]
    django_template_dirs =
      /my/project/template/directory
      /my/project/template/directory2

In buildout.cfg::

    [buildout]
    parts = django
    # relative to the buildout directory
    django_template_dirs +=
      templates
    eggs =
      egg1
      egg2

    [django]
    recipe = djangorecipe
    eggs = ${buildout:eggs}
    project = myproject
    template = mytemplate

The template directories are explored in this order (reverse of the order in
which they are defined to enable overriding):

1. buildout/directory/templates
2. /my/project/template/directory2
3. /my/project/template/directory



CHANGES.rst
Changes
=======

1.8 (unreleased)
----------------

- Tests can now be run on windows
- The project directory can now be set as the buildout directory by setting
  the 'project' option to a single dot (.)
- The project is now generated from template directories that can be
  user-defined, using python's built-in template engine for string
  substitution


1.7 (2013-12-11)
----------------

- Add option to change the wsgi script name. Thanks hedleyroos (Fixes pull #74)

1.6 (2013-10-28)
----------------

- Djangorecipe works with django 1.6 now.

- Tested with Django 1.4, 1.5 and 1.6. Pre-1.4 support is gone, now. Also
  tested on Python 2.6/2.7, 3.2/3.3.

- Moved to buildout 2 support only.

- Removed generation of fastcgi script. You can run it easily as ``bin/django
  runfcgi`` and it will be deprecated in Django 1.7 anyway.


1.5 (2013-01-25)
----------------

- Removed support for a different python version than the one you use to run
  buildout with. Previously, you could run your buildout with 2.6 but get
  Django to use 2.7 instead. zc.buildout 2.0 doesn't allow it anymore, so we
  removed it too.


1.4 (2013-01-15)
----------------

- Added initialization code support. Thanks to anshumanb, jjmurre. (Closes #58).


1.3 (2012-09-07)
----------------

- Removed deprecation warning in Django 1.4. Fixes #49, thanks Shagi.

- Added documentation for use with mr.developer. Thanks shagi (closes issue #45)

- Added Travis support.


1.2.1 (2012-05-15)
------------------

- Fixed broken 1.2 release (missing ``*.rst`` files due to a recent txt-to-rst
  rename action).


1.2 (2012-05-14)
----------------

- Removed location path from recipe. Thanks bleskes (fixes issue #50).

1.1.2
-----

- Added correct url to the deprecation warning

1.1.1
-----

- Fixed Python3 Trove classifiers

1.1
---

- Support python3.
- Changed buildout and the tests to run the tests under nose.
- Removed some old pre 0.99 unittests that dealt with download support.

1.0
---

- Stable release with a real 1.0 version.
- Made djangorecipe more pep08 compliant.

0.99
----

- Djangorecipe now depends on Django. The use of the `version =` statement
  is deprecated. Specify the django version in the
  `[versions]` section. Install django via mr.developer if you need to use
  an svn/git/hg repository. For other uses
  Versionpin djangorecipe to 0.23.1 if you don't want to upgrade.
  Thanks to Reinout van Rees for help with this release.

- Removed subversion and download support.

0.23.1
------

- Added a missing 'import os'

0.23
----

- Support for settings/urls boilerplate for django 1.2 and django 1.3.
  It defaults to 1.3 when the version isn't 1.2.

0.22
----

- Added support for svn urls with spaces. Thanks to Brad103 (fixes #537718).

- Updated code and buildout to use newest zc.recipe.egg,
  zc.recipe.testrunner and python-dateutil.

0.21
----

- The admin url is now configured for django 1.1 or higher. Thanks to
  Sam Charrington (fixes #672220).

- Bootstrap.py updated (fixes #501954).

0.20
----

- The recipe know makes the `django` package know to setuptools during install.
  This closes #397864. Thanks to Daniel Bruce and Dan Fairs for the patch.

- Fixed #451065 which fixes a problem with the WSGI log file option.

- Added the posibilty to configure more FCGI related settings. Thanks to Vasily
  Sulatskov for the patch.

0.19.2
------

- The generated WSGI & FCGI scripts are now properly removed when
  options change (fixes #328182). Thanks to Horst Gutmann for the
  patch.

- Scripts are now updated when dependencies change. This fixes #44658,
  thanks to Paul Carduner for the patch.

0.19.1
------

- Applied fix for the change in WSGI script generation. The previous
  release did not work properly.

0.19
----

- When running again with non-newest set the recipe will no longer
  update the Subversion checkout. Thanks to vinilios for the patch.

- The WSGI and FCGI scripts are now generated using Buildout's own
  system. This makes them more similar to the generated manage script
  with regard to the setup of paths. Thanks to Jannis Leidel for the
  patch.

0.18
----

- Paths from eggs and extra-paths now get precedence over the default
  system path (fixes #370420). Thanks to Horst Gutmann for the patch.

- The generated WSGI script now uses the `python` option if
  present. This fixes #361695.

0.17.4
------

- Fixed a problem when not running in verbose mode (fixes #375151).

0.17.3
------

- Removed dependency on setuptools_bzr since it does not seem to work
  like I expected.

0.17.2
------

- Changed the download code to use urllib2. This should make it work
  from behind proxies (fixes #362822). Thanks to pauld for the patch.

0.17.1
------

- Fixed a problem with the new WSGI logging option #348797. Thanks to
  Bertrand Mathieu for the patch.

- Disable generation of the WSGI log if "wsgilog" isn't set, thanks to
  Jacob Kaplan-Moss for the patch.

- Updated buildout.cfg and .bzrignore, thanks Jacob Kaplan-Moss.

0.17
----

- Added an option to specify a log file for output redirection from
  the WSGI script. Thanks to Guido Wesdorp for the patch.

0.16
----

- Subversion aliases are now supported (something like
  svn+mystuff://myjunk). Thanks to Remco for the patch.

0.15.2
------

- Update to move pth-files finder from the __init__ method to the
  install method so it runs in buildout-order, else it looks for pth
  files in dirs that may not yet exist. Thanks to Chris Shenton for
  the update to his original patch.

0.15.1
------

- Update to make the previously added pth-files option better
  documented.

0.15
----

- Added "pth-files" option to add libraries to extra-paths from
  site .pth files. Thanks to Chris Shenton for the patch.

0.14
----

- The recipe now supports creating a FCGI script. Thanks to Jannis
  Leidel for the patch.

- When downloading a Django recipe for the first time the recipe now
  properly reports the url it is downloading from.

0.13
----


src/djangorecipe/recipe.py
from random import choice
import os
import logging
import re
import sys
import shutil
from datetime import date
from distutils.version import StrictVersion

from zc.buildout import UserError
import zc.recipe.egg

from djangorecipe.templating import process, process_tree, script_template


class Recipe(object):
    def __init__(self, buildout, name, options):
        # The use of version is deprecated.
        if 'version' in options:
            raise UserError('The version option is deprecated. '
                            'Read about the change on '
                            'http://pypi.python.org/pypi/djangorecipe/0.99')
        self.log = logging.getLogger(name)
        self.egg = zc.recipe.egg.Egg(buildout, options['recipe'], options)

        self.buildout, self.name, self.options = buildout, name, options
        options['location'] = os.path.join(
            buildout['buildout']['parts-directory'], name)
        options['bin-directory'] = buildout['buildout']['bin-directory']

        options.setdefault('project', 'project')
        options.setdefault('settings', 'development')

        # gets the root package path
        self.get_root_pkg()

        options.setdefault('urlconf', self.root_pkg + 'urls')
        options.setdefault(
            'media_root',
            "os.path.join(os.path.dirname(__file__), 'media')")
        # Set this so the rest of the recipe can expect the values to be
        # there. We need to make sure that both pythonpath and extra-paths are
        # set for BBB.
        if 'extra-paths' in options:
            options['pythonpath'] = options['extra-paths']
        else:
            options.setdefault('extra-paths', options.get('pythonpath', ''))

        options.setdefault('initialization', '')

        # mod_wsgi support script
        options.setdefault('wsgi', 'false')
        options.setdefault('wsgilog', '')
        options.setdefault('logfile', '')

    def install(self):
        base_dir = self.buildout['buildout']['directory']

        if self.root_pkg:
            project_dir = os.path.join(base_dir, self.options['project'])
        else:
            project_dir = base_dir

        extra_paths = self.get_extra_paths()
        requirements, ws = self.egg.working_set(['djangorecipe'])

        script_paths = []
        # Create the Django management script
        script_paths.extend(self.create_manage_script(extra_paths, ws))

        # Create the test runner
        script_paths.extend(self.create_test_runner(extra_paths, ws))

        # Make the wsgi and fastcgi scripts if enabled
        script_paths.extend(self.make_scripts(extra_paths, ws))

        # Create default project files if we haven't got a project
        # egg specified, and if the settings don't already exist
        if not self.options.get('projectegg'):
            settings_path = \
                os.path.join(project_dir, *self.options['settings'].split('.'))
            if not (os.path.exists(os.path.join(settings_path, '__init__.py'))
                    or os.path.exists('.'.join((settings_path, 'py')))):
                print settings_path
                self.create_project(project_dir)
            else:
                self.log.debug(
                    'Skipping creating project files for %(project)s since '
                    'its main settings module exists' % self.options)

        return script_paths

    def create_manage_script(self, extra_paths, ws):
        return zc.buildout.easy_install.scripts(
            [(self.options.get('control-script', self.name),
              'djangorecipe.manage', 'main')],
            ws, sys.executable, self.options['bin-directory'],
            extra_paths=extra_paths,
            arguments="'%s%s'" % (self.root_pkg, self.options['settings']),
            initialization=self.options['initialization'])

    def create_test_runner(self, extra_paths, working_set):
        apps = self.options.get('test', '').split()
        # Only create the testrunner if the user requests it
        if apps:
            return zc.buildout.easy_install.scripts(
                [(self.options.get('testrunner', 'test'),
                  'djangorecipe.test', 'main')],
                working_set, sys.executable,
                self.options['bin-directory'],
                extra_paths=extra_paths,
                arguments="'%s%s', %s" % (
                    self.root_pkg,
                    self.options['settings'],
                    ', '.join(["'%s'" % app for app in apps])),
                initialization=self.options['initialization'])
        else:
            return []

    def create_project(self, project_dir):
        # create the project directory if it does not exist
        if not os.path.exists(project_dir):
            os.makedirs(project_dir)

        # retrieve user-provided template directories
        template_dirs = self.buildout['djangorecipe'] \
                            .get('template-dirs', '') \
                            .splitlines()

        # retrieve template name
        template_name = self.options.get('template', None)

        if template_name and template_dirs:
            # the user provided a template to load

            # look for a template in the template directories provided
            # in reverse so that the last setting is prioritary
            for d in reversed(template_dirs):
                d = os.path.abspath(d)
                if template_name in os.listdir(d):
                    # we have a template candidate, load it
                    temp_path = os.path.join(d, template_name)

        else:
            # no template name was provided
            temp_path = os.path.join(os.path.dirname(__file__),
                                     'templates')

            # Find the current Django versions in the buildout versions.
            version = None
            b_versions = self.buildout.get('versions')
            if b_versions:
                django_version = (
                    b_versions.get('django') or
                    b_versions.get('Django')
                )
                if django_version:
                    version_re = re.compile("\d+\.\d+")
                    match = version_re.match(django_version)
                    version = match and match.group()

                if version not in os.listdir(temp_path):
                    # the version is invalid (no template)
                    version = None

            # if the version could not be found, gets the latest one from the
            # default templates names
            if not version:
                av_versions = os.listdir(temp_path)
                av_versions.sort(key=StrictVersion)
                version = av_versions[-1]

            temp_path = os.path.join(temp_path, version)

        # prepare templating engine
        template_vars = self.get_template_vars()
        template_vars.update(self.options)

        # copy files and run templating engine
        for sub in os.listdir(temp_path):
            src_path = os.path.join(temp_path, sub)
            tgt_path = os.path.join(project_dir, sub)
            if os.path.exists(tgt_path):
                sys.stderr.write('ERROR: %s already exists in %s and ' \
                    'cannot be overwritten by djangorecipe\'s template ' \
                    'engine.\n' % (sub, project_dir))
            else:
                if os.path.isdir(src_path):
                    # copy the subdirectory tree
                    shutil.copytree(src_path, tgt_path)
                    process_tree(tgt_path, template_vars)
                else:
                    # copy the file and run templating engine
                    shutil.copy(src_path, tgt_path)
                    process(tgt_path, template_vars)

    def make_scripts(self, extra_paths, ws):
        scripts = []
        protocol = 'wsgi'

        if self.options.get(protocol, '').lower() == 'true':
            _script_template = zc.buildout.easy_install.script_template
            protocol = 'wsgi'
            zc.buildout.easy_install.script_template = \
                zc.buildout.easy_install.script_header + \
                script_template[protocol]

            scripts.extend(
                zc.buildout.easy_install.scripts(
                    [(self.options.get('wsgi-script') or
                      '%s.%s' % (self.options.get('control-script',
                                                  self.name),
                                 protocol),
                      'djangorecipe.%s' % protocol, 'main')],
                    ws,
                    sys.executable,
                    self.options['bin-directory'],
                    extra_paths=extra_paths,
                    arguments="'%s%s', logfile='%s'" % (
                        self.root_pkg, self.options['settings'],
                        self.options.get('logfile')),
                    initialization=self.options['initialization']))
            zc.buildout.easy_install.script_template = _script_template

        return scripts

    def get_template_vars(self):
        today = date.today()
        t_vars = {
            'secret': self.generate_secret(),
            'project_name': self.project_name,
            'root_pkg': self.root_pkg,
            'year': today.year,
            'month': today.month,
            'day': today.day
        }
        t_vars.update(self.options)
        t_vars.update(self.buildout.get('djangorecipe', {}))
        return t_vars

    def get_root_pkg(self):
        project = self.options.get('projectegg', self.options['project'])
        if project == '.':
            self.root_pkg = ''
            self.project_name = os.path.basename(
                os.path.dirname(self.buildout['buildout']['directory'])
            )
        else:
            self.root_pkg = project + '.'
            self.project_name = project

    def get_extra_paths(self):
        extra_paths = [self.buildout['buildout']['directory']]

        # Add libraries found by a site .pth files to our extra-paths.
        if 'pth-files' in self.options:
            import site
            for pth_file in self.options['pth-files'].splitlines():
                pth_libs = site.addsitedir(pth_file, set())
                if not pth_libs:
                    self.log.warning(
                        "No site *.pth libraries found for pth_file=%s" % (
                            pth_file,))
                else:
                    self.log.info("Adding *.pth libraries=%s" % pth_libs)
                    self.options['extra-paths'] += '\n' + '\n'.join(pth_libs)

        pythonpath = [p.replace('/', os.path.sep) for p in
                      self.options['extra-paths'].splitlines() if p.strip()]

        extra_paths.extend(pythonpath)
        return extra_paths

    def update(self):
        extra_paths = self.get_extra_paths()
        requirements, ws = self.egg.working_set(['djangorecipe'])
        # Create the Django management script
        self.create_manage_script(extra_paths, ws)

        # Create the test runner
        self.create_test_runner(extra_paths, ws)

        # Make the wsgi and fastcgi scripts if enabled
        self.make_scripts(extra_paths, ws)

    def generate_secret(self):
        chars = 'abcdefghijklmnopqrstuvwxyz0123456789!@#$%^&*(-_=+)'
        return ''.join([choice(chars) for i in range(50)])
